File: 0105 - Construct Binary Tree from Preorder and Inorder Traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    TreeNode* build(vector<int>& preorder, int& idx, vector<int>& inorder, int l, int r, unordered_map<int, int>& m){
        if(l > r) return nullptr;
        int mid = m[preorder[idx]];
        TreeNode *res = new TreeNode(preorder[idx++]);
        res->left = build(preorder, idx, inorder, l, mid-1, m);
        res->right = build(preorder, idx, inorder, mid+1, r, m);
        return res;
    }
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int n = inorder.size(), idx = 0;
        unordered_map<int, int> m;
        for(int i = 0; i < n; i++) m[inorder[i]] = i;
        return build(preorder, idx, inorder, 0, n-1, m);
    }
};

File: 0105 - Construct Binary Tree from Preorder and Inorder Traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if inorder:
            idx = inorder.index(preorder.pop(0))
            res = TreeNode(inorder[idx])
            res.left = self.buildTree(preorder, inorder[:idx])
            res.right = self.buildTree(preorder, inorder[idx+1:])
            return res
        return None